# Flask
---

* Es un micro framework para Python
* Permite crear aplicaciones web en forma rápida
* [Guia de usuario](https://flask.palletsprojects.com/es/stable/)

## Preparación del entorno de programación

### 1. Instalar Python

[python](https://www.python.org/downloads/)

### 2. Crear un directorio

```shell
mkdir prueba
```

### 3. Entrar al directorio

```shell
cd prueba
```

### 4. Crear entorno virtual

```shell
python -m venv venv 
```

### 5. Activar entorno virtual

```shell
venv\Scripts\activate
```

### 6. Instalar Flask

```shell
pip install flask
```

### 7. Verificar instalación

```shell
pip list
```

### 8. Crear archivo hello.py

```shell
touch hello.py
```

## Aplicación Hello World

In [1]:
%%file hello.py
from flask import Flask

app = Flask(__name__)

@app.route('/')
def hello_world():
    return 'Hello, World!'

Overwriting hello.py


* Toda aplicación Flask es una **instancia de Flask**
* **\_\_name__**: nombre del módulo, para que Flask sepa dónde encontrar los recursos
* Flask se encarga de hacer transparente el cómo a partir de una petición URL se ejecuta finalmente la función
* El decorador **route** de la aplicación app es el encargado de decirle a Flask qué URL debe ejecutar su correspondiente función
* La función debe **devolver la respuesta** que será mostrada en el navegador del usuario
* El tipo de contenido por **defecto es HTML**

## Ejecución del servidor

* Flask viene con un servidor interno

### 1. Indicar al servidor que aplicación debe lanzar mediante declaración de variable de entorno

In [2]:
!export FLASK_APP=hello.py

### 2. Ejecutar el servidor

In [3]:
!flask --app hello run

 * Serving Flask app 'hello'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 20:30:32] "GET / HTTP/1.1" 200 -
^C


* Para comprobar que la aplicación funciona, se entra al navegador y en la barra de direcciones introducir **localhost:5000/**
* Para cambiar el puerto por cualquier motivo se puede hacer de dos formas distintas:
  * Estableciendo la variable de entorno **FLASK_RUN_PORT** en un puerto diferente
  * Indicando el puerto al lanzar el servidor: **flask run --port 6000**
* Para aceptar peticiones de otras computadoras de nuestra red se lanza el servidor de la siguiente manera: **flask run --host 0.0.0.0**
* Para salir del servidor **CTRL + C**

## Aplicación con formato

In [1]:
%%file formato.py
from flask import Flask

app = Flask(__name__)

@app.route('/contacto')
def contacto():
    return '<h1>Contacto</h1>'

Overwriting formato.py


In [2]:
!export FLASK_APP=formato.py

In [3]:
!flask --app formato run

 * Serving Flask app 'formato'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 19:51:30] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 19:51:37] "GET /contacto HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 19:51:42] "GET /.well-known/appspecific/com.chrome.devtools.json HTTP/1.1" 404 -
^C


## Escape HTML

In [1]:
%%file param_esc.py
from flask import Flask
from markupsafe import escape

app = Flask(__name__)

@app.route('/<name>')
def hello(name):
    return f'Hello {escape(name)}'

Overwriting param_esc.py


In [2]:
!export FLASK_APP=param_esc.py

In [3]:
!flask --app param_esc run

 * Serving Flask app 'param_esc'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 20:35:21] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:35:27] "GET /hola/pablo HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:35:30] "GET /pepe HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 20:35:57] "GET /<script>alert('pepe')</script> HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:36:15] "GET /<script>alert('pepe') HTTP/1.1" 200 -
^C


* Cualquier valor proporcionado por el usuario debe ser escapado para protegerlo de **ataques de inyección**. Por ej: \<br>\<br>Pepe
* El escape hace que se **renderice como texto**, en lugar de ejecutar el script en el navegador del usuario

## Enrutamiento

* Las aplicaciones web modernas utilizan URLs significativas para ayudar a los usuarios que puedan recordar y utilizar para visitar directamente una página
* Decorador **route()** para vincular una función a una URL
* También se puede hacer que **partes de la URL sean dinámicas** y adjuntar múltiples reglas a una función

In [1]:
%%file enrutamiento.py
from flask import Flask

app = Flask(__name__)

@app.route('/')
def index():
    return 'Index Page'

@app.route('/hello')
def hello():
    return 'Hello, World'

Writing enrutamiento.py


In [2]:
!export FLASK_APP=enrutamiento.py

In [3]:
!flask --app enrutamiento run

 * Serving Flask app 'enrutamiento'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 20:24:04] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 20:24:08] "GET /hola/pablo HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:24:10] "GET /hello HTTP/1.1" 200 -
^C


## Normas variables

* En una aplicación web no todas las páginas tienen una **URL definida de antemano**, como es el caso de la página principal
* A una URL se le pueden agregar **secciones variables o parametrizadas con "param"**
* La vista recibirá "param" como un parámetro con ese mismo nombre
* Opcionalmente se puede indicar un conversor "converter" para especificar el tipo de dicho parámetro así "converter:param"

In [1]:
%%file param.py
from flask import Flask
from markupsafe import escape

app = Flask(__name__)

@app.route('/user/<username>')
def show_user_profile(username):
    return f'User {escape(username)}'

@app.route('/post/<int:post_id>')
def show_post(post_id):
    return f'Post {post_id}'

@app.route('/path/<path:subpath>')
def show_subpath(subpath):
    return f'Subpath {escape(subpath)}'

Overwriting param.py


In [2]:
!export FLASK_APP=param.py

In [3]:
!flask --app param run

 * Serving Flask app 'param'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 20:42:53] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:43:02] "GET /user/pepe HTTP/1.1" 200 -
^C


* En un navegador: localhost:5000/user/pepe

* Por defecto, en Flask existen los siguientes conversores:

|            |                                                                                                           |
| --         | --                                                                                                        |
| **string** | (Por defecto) acepta cualquier texto sin barra                                                            |
| **int**    | acepta números enteros positivos                                                                          |
| **float**  | acepta valores positivos en coma flotante                                                                 |
| **path**   | como string pero también acepta barras inclinadas                                                         |
| **uuid**   | acepta cadenas con formato UUID (32 dígitos hexadecimales). Ejemplo: 550e8400-e29b-41d4-a716-446655440000 |

## URLs únicas / Comportamiento de la redirección

* Las dos reglas siguientes difieren en el uso de la barra diagonal final:

In [1]:
%%file urls.py
from flask import Flask
from markupsafe import escape

app = Flask(__name__)

@app.route('/projects/')
def projects():
    return 'The project page'

@app.route('/about')
def about():
    return 'The about page'

Writing urls.py


In [2]:
!export FLASK_APP=urls.py

In [3]:
!flask --app urls run

 * Serving Flask app 'urls'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 20:49:16] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 20:56:41] "GET /projects HTTP/1.1" 308 -
127.0.0.1 - - [04/Apr/2026 20:56:41] "GET /projects/ HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 20:56:50] "GET /about HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 20:56:54] "GET /about/ HTTP/1.1" 404 -
^C


| La URL canónica del endpoint ... |                             | Si se accede a la URL:                          |
| --                               | --                          | --                                              |
| **/projects**                    | tiene una barra al final    | **/projects**: Flask redirige a la URL canónica |
| **/about**                       | no tiene una barra al final | **/about/**: produce un error 404 "Not Found"   |

* Esto ayuda a mantener las URLs únicas para estos recursos, lo que ayuda a los motores de búsqueda a evitar indexar la misma página dos veces

## Métodos HTTP

* Las aplicaciones web utilizan diferentes métodos HTTP para acceder a las URL
* Por defecto, una ruta sólo responde a peticiones GET
* El ejemplo mantiene todos los métodos para la ruta dentro de una función, lo que puede ser útil si cada parte utiliza algunos datos comunes

### Diferentes métodos misma función

In [1]:
%%file methods.py
from flask import Flask
from flask import request

app = Flask(__name__)

@app.route('/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        return 'post'
    return 'get'

Writing methods.py


In [2]:
!export FLASK_APP=methods.py

In [3]:
!flask --app methods run

 * Serving Flask app 'methods'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 21:04:25] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:04:33] "GET /login HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 21:06:22] "POST /login HTTP/1.1" 200 -
^C


### Diferentes métodos en diferentes funciones

In [5]:
%%file http_methods.py
from flask import Flask

app = Flask(__name__)

@app.get('/login')
def login_get():
    return 'get'

@app.post('/login')
def login_post():
    return 'post'

Overwriting http_methods.py


In [8]:
!export FLASK_APP=http_methods.py

In [9]:
!flask --app http_methods run

 * Serving Flask app 'http_methods'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 21:12:03] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:12:08] "GET /login HTTP/1.1" 200 -
^C


## Template sin parámetro

* Uso de plantillas para la generación de contenido HTML dinámico
* En Flask, por convención, a las funciones que están asociadas a una URL se les llama "vistas"
1. Crear dentro del proyecto una carpeta llamada **templates**, Flask buscará las plantillas dentro de este directorio
2. Crear dentro de la carpeta **templates** un archivo html. Ej. index.html

In [4]:
%%file templates/index.html
<html>
<head>
    <title> Flask App </title>
</head>
<body>
    <h1>Hello</h1>
    <h2>World!</h2>
</body>
</html>

Overwriting templates/index.html


In [5]:
%%file template_sin_param.py
from flask import Flask, render_template

app = Flask(__name__)

@app.route('/home')
def home():
    return render_template('index.html')

Overwriting template_sin_param.py


In [6]:
!export FLASK_APP=template_sin_param.py

In [7]:
!flask --app template_sin_param run

 * Serving Flask app 'template_sin_param'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 21:24:30] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:24:36] "GET /hola/pablo HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:24:50] "GET /hola/pablo HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:24:51] "GET /home HTTP/1.1" 200 -
^C


## Template con parámetro

In [1]:
%%file templates/index2.html
<html>
<head>
    <title> Flask App </title>
</head>
<body>
    <h1>Hello {{name}}</h1>
</body>
</html>

Overwriting templates/index2.html


* En el archivo html, se utiliza una sintaxis llamada jinja2 (igual que en django)
* Una plantilla jinja2 no es más que un archivo que contiene datos estáticos junto con bloques para generar contenido dinámico
* El resultado de renderizar una plantilla es un documento HTML en el que los bloques de generación de contenido dinámico han sido procesados

In [2]:
%%file template_con_param.py
from flask import Flask, render_template

app = Flask(__name__)

@app.route('/home')
def home():
    return render_template('index2.html', name='mundo')

Overwriting template_con_param.py


In [3]:
!export FLASK_APP=template_con_param.py

In [4]:
!flask --app template_con_param run

 * Serving Flask app 'template_con_param'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 21:27:00] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:27:03] "GET /hola/pablo HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:27:04] "GET /home HTTP/1.1" 200 -
^C


## Template con parámetro mediante param

In [1]:
%%file templates/index3.html
<html>
<head>
    <title> Flask App </title>
</head>
<body>
    {% if person %}
         <h1>Hello {{ person }}!</h1>
    {% else %}
         <h1>Hello, World!</h1>
    {% endif %}
</body>
</html>

Overwriting templates/index3.html


In [2]:
%%file template_con_param_param.py
from flask import Flask, render_template

app = Flask(__name__)

@app.route('/home')
@app.route('/home/<name>')
def hello(name=None):
    return render_template('index3.html', person=name)

Writing template_con_param_param.py


In [3]:
!export FLASK_APP=template_con_param_param.py

In [ ]:
!flask --app template_con_param_param run

 * Serving Flask app 'template_con_param_param'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [04/Apr/2026 21:34:34] "GET / HTTP/1.1" 404 -
127.0.0.1 - - [04/Apr/2026 21:34:37] "GET /home HTTP/1.1" 200 -
127.0.0.1 - - [04/Apr/2026 21:34:43] "GET /home/pepe HTTP/1.1" 200 -


## Archivos estáticos

* Su contenido no cambia a lo largo del ciclo de ejecución de una aplicación web
* Ejemplo: archivos CSS, imágenes, código javascript
* Flask facilita una vista para servir los recursos estáticos de la aplicación, evitando de preocuparse de configuraciones
* Estos archivos se deben ubicar en un directorio llamado **static** situado al mismo nivel que el directorio **template**. Este directorio estará accesible en la URL **/static**

## Uso de formularios en Flask

* A través de los formularios los usuarios pueden interactuar con la aplicación enviando cualquier tipo de información al servidor
* El protocolo HTTP define una serie de de métodos de petición para indicar la acción que se desea realizar sobre un recurso determinado:
  * GET: debe emplearse para solicitar un recurso (recuperar datos)
  * POST: se utiliza para enviar una entidad a un recurso específico, de manera que sen envian datos al servidor donde son procesados

In [28]:
%%file templates/layout.html
<html>
<head>
    <title> Flask App </title>
</head>
<body>
    {% block body %}
    {% endblock %}
</body>
</html>

Overwriting templates/layout.html


In [29]:
%%file templates/index3.html
{% extends 'layout.html' %}
{% block body %}
<h1>Bienvenido</h1>
<form action="{{ url_for('contacto') }}" method='post'>
    <input type='text' name='nombre' placeholder='Ingresa un nombre'>
    <button>Enviar</button>
</form>
{% endblock %}

Overwriting templates/index3.html


In [30]:
%%file templates/contacto.html
{% extends 'layout.html' %}
{% block body %}
<h1>Hola {{nombre}}</h1>
{% endblock %}

Overwriting templates/contacto.html


In [1]:
%%file form.py
from flask import Flask, render_template, request

app = Flask(__name__)

@app.route('/')
def home():
    return render_template('index3.html')

@app.route('/contacto', methods=['POST'])   # Por default es 'GET'
def contacto():
    nombre = request.form.get('nombre')     # 'nombre' es el valor del attr 'name' del input
    return render_template('contacto.html', nombre=nombre)

Overwriting form.py


In [2]:
!export FLASK_APP=form.py

In [3]:
!flask --app form run

 * Serving Flask app 'form'
 * Debug mode: off
 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
127.0.0.1 - - [30/Aug/2025 10:54:44] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [30/Aug/2025 10:54:50] "POST /contacto HTTP/1.1" 200 -
^C


## Objeto request